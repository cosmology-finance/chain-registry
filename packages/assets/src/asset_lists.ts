import { AssetList } from '@chain-registry/types';

import _mainnet from './mainnet/assets';
import _testnet from './testnet/assets';
import _noncosmos from './noncosmos/assets'

const asset_lists: AssetList[] = [..._mainnet,..._testnet,..._noncosmos];

export default asset_lists;
