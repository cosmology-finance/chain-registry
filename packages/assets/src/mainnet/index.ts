export * as _8ball from './8ball';
export * as acrechain from './acrechain';
export * as agoric from './agoric';
export * as aioz from './aioz';
export * as akash from './akash';
export * as andromeda from './andromeda';
export * as archway from './archway';
export * as arkh from './arkh';
export * as assetmantle from './assetmantle';
export * as aura from './aura';
export * as axelar from './axelar';
export * as bandchain from './bandchain';
export * as beezee from './beezee';
export * as bitcanna from './bitcanna';
export * as bitsong from './bitsong';
export * as bluzelle from './bluzelle';
export * as bostrom from './bostrom';
export * as canto from './canto';
export * as carbon from './carbon';
export * as celestia from './celestia';
export * as cerberus from './cerberus';
export * as chain4energy from './chain4energy';
export * as cheqd from './cheqd';
export * as chihuahua from './chihuahua';
export * as cifer from './cifer';
export * as comdex from './comdex';
export * as composable from './composable';
export * as coreum from './coreum';
export * as cosmoshub from './cosmoshub';
export * as crescent from './crescent';
export * as cronos from './cronos';
export * as cryptoorgchain from './cryptoorgchain';
export * as cudos from './cudos';
export * as decentr from './decentr';
export * as desmos from './desmos';
export * as dig from './dig';
export * as doravota from './doravota';
export * as dydx from './dydx';
export * as dymension from './dymension';
export * as dyson from './dyson';
export * as echelon from './echelon';
export * as emoney from './emoney';
export * as empowerchain from './empowerchain';
export * as evmos from './evmos';
export * as fetchhub from './fetchhub';
export * as furya from './furya';
export * as fxcore from './fxcore';
export * as galaxy from './galaxy';
export * as gateway from './gateway';
export * as genesisl1 from './genesisl1';
export * as gitopia from './gitopia';
export * as gravitybridge from './gravitybridge';
export * as haqq from './haqq';
export * as humans from './humans';
export * as impacthub from './impacthub';
export * as imversed from './imversed';
export * as injective from './injective';
export * as irisnet from './irisnet';
export * as jackal from './jackal';
export * as juno from './juno';
export * as kava from './kava';
export * as kichain from './kichain';
export * as konstellation from './konstellation';
export * as kujira from './kujira';
export * as kyve from './kyve';
export * as lambda from './lambda';
export * as likecoin from './likecoin';
export * as lumenx from './lumenx';
export * as lumnetwork from './lumnetwork';
export * as mars from './mars';
export * as medasdigital from './medasdigital';
export * as meme from './meme';
export * as microtick from './microtick';
export * as migaloo from './migaloo';
export * as neutron from './neutron';
export * as nibiru from './nibiru';
export * as noble from './noble';
export * as nois from './nois';
export * as nolus from './nolus';
export * as nomic from './nomic';
export * as nyx from './nyx';
export * as odin from './odin';
export * as okexchain from './okexchain';
export * as omniflixhub from './omniflixhub';
export * as onomy from './onomy';
export * as oraichain from './oraichain';
export * as osmosis from './osmosis';
export * as panacea from './panacea';
export * as passage from './passage';
export * as persistence from './persistence';
export * as planq from './planq';
export * as point from './point';
export * as provenance from './provenance';
export * as pundix from './pundix';
export * as pylons from './pylons';
export * as quasar from './quasar';
export * as quicksilver from './quicksilver';
export * as qwoyn from './qwoyn';
export * as realio from './realio';
export * as rebus from './rebus';
export * as regen from './regen';
export * as rizon from './rizon';
export * as scorum from './scorum';
export * as secretnetwork from './secretnetwork';
export * as sei from './sei';
export * as sentinel from './sentinel';
export * as sge from './sge';
export * as shareledger from './shareledger';
export * as shentu from './shentu';
export * as sifchain from './sifchain';
export * as sommelier from './sommelier';
export * as source from './source';
export * as stafihub from './stafihub';
export * as stargaze from './stargaze';
export * as starname from './starname';
export * as stride from './stride';
export * as teritori from './teritori';
export * as terra from './terra';
export * as terra2 from './terra2';
export * as tgrade from './tgrade';
export * as umee from './umee';
export * as unification from './unification';
export * as uptick from './uptick';
export * as vidulum from './vidulum';
export * as xpla from './xpla';
