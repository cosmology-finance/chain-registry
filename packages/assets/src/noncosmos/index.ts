export * as picasso from './picasso';
export * as composablepolkadot from './composablepolkadot'