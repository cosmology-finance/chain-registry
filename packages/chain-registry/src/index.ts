export * from './mainnet';
export * from './testnet';
export * from './noncosmos';

import all from './all';

export default all;

const { assets, chains, ibc } = all;

export { assets, chains, ibc };
