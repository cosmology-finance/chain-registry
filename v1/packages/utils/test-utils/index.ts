import { AssetList, Chain, IBCInfo } from '@chain-registry/types';

import _assets from '../../../../__fixtures__/v1/assets.json';
import _chains from '../../../../__fixtures__/v1/chains.json';
import _ibc from '../../../../__fixtures__/v1/ibc.json';

export const assets: AssetList[] = _assets as AssetList[];
// TODO update fixtures, they may be missing properties and/or old schema
export const chains: Chain[] = _chains as any;

export const ibc: IBCInfo[] = _ibc as IBCInfo[];
