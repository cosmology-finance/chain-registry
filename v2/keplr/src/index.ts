import { Asset, AssetList, Chain } from '@chain-registry/types';
import { Bech32Address } from '@keplr-wallet/cosmos';
import { ChainInfo, Currency, FeeCurrency } from '@keplr-wallet/types';
import semver from 'semver';

const getRpc = (chain: Chain): string => chain.apis?.rpc?.[0]?.address ?? '';
const getRest = (chain: Chain): string => chain.apis?.rest?.[0]?.address ?? '';
const getExplr = (chain: Chain): string => chain.explorers?.[0]?.url ?? '';

const cleanVer = (ver: string) => {
  if (!semver.valid(ver)) {
    // try to split by @ for repo@version
    ver = ver.split('@').pop();
    if (semver.valid(ver)) return ver;

    const spaces = ver.split('.').length;
    switch (spaces) {
      case 1:
        return ver + '.0.0';
      case 2:
        return ver + '.0';
      case 3:
      default:
        throw new Error(
          'contact @chain-registry/keplr maintainers: bad version'
        );
    }
  }
};

export const chainRegistryChainToKeplr = (
  chain: Chain,
  assets: AssetList[],
  options: {
    getRpcEndpoint?: (chain: Chain) => string;
    getRestEndpoint?: (chain: Chain) => string;
    getExplorer?: (chain: Chain) => string;
  } = {
    getRpcEndpoint: getRpc,
    getRestEndpoint: getRest,
    getExplorer: getExplr
  }
): ChainInfo => {
  if (!options.getRestEndpoint) options.getRestEndpoint = getRest;
  if (!options.getRpcEndpoint) options.getRpcEndpoint = getRpc;
  if (!options.getExplorer) options.getExplorer = getExplr;

  const features = [];
  // if NOT specified, we assume stargate, sorry not sorry
  const sdkVer = chain.codebase?.cosmos_sdk_version
    ? cleanVer(chain.codebase?.cosmos_sdk_version)
    : '0.40';

  // stargate
  if (semver.satisfies(sdkVer, '>=0.40')) features.push('stargate');
  // no-legacy-stdTx
  if (semver.satisfies(sdkVer, '>=0.43')) features.push('no-legacy-stdTx');
  // until further notice, assume 'ibc-transfer'
  features.push('ibc-transfer');

  // ibc-go
  if (semver.satisfies(sdkVer, '>=0.45')) features.push('ibc-go');

  if (chain.codebase?.cosmwasm_enabled) {
    features.push('cosmwasm');
    const wasmVer = cleanVer(chain.codebase.cosmwasm_version ?? '0.24');
    if (semver.satisfies(wasmVer, '>=0.24')) features.push('wasmd_0.24+');
  }

  const chainAssets =
    assets.find((asset) => asset.chain_name === chain.chain_name)?.assets || [];

  const feeDenoms =
    chain.fees?.fee_tokens.map<string>((feeToken) => feeToken.denom) || [];

  /**
   * FROM KEPLR chain-info.d.ts:
   * This is used to set the fee of the transaction.
   * If this field is empty, it just use the default gas price step (low: 0.01, average: 0.025, high: 0.04).
   * And, set field's type as primitive number because it is hard to restore the prototype after deserialzing if field's type is `Dec`.
   */
  interface GasPriceStep {
    low: number;
    average: number;
    high: number;
  }

  const gasPriceSteps: Record<string, GasPriceStep> = chain.fees?.fee_tokens?.reduce((m, feeToken) => {
    m[feeToken.denom] = {
      low: feeToken.low_gas_price ?? 0.01,
      average: feeToken.average_gas_price ?? 0.025,
      high: feeToken.high_gas_price ?? 0.04
    };
    return m;
  }, {} as Record<string, GasPriceStep>);

  const stakingDenoms =
    chain.staking?.staking_tokens.map<string>(
      (stakingToken) => stakingToken.denom
    ) || [];

  const currencies: Currency[] = chainAssets.map((currency: Asset) => {
    return {
      coinDenom: currency.symbol,
      coinMinimalDenom: currency.base,
      coinDecimals: currency.denom_units.filter(
        (denomUnit: { denom: string }) => denomUnit.denom === currency.display
      )[0]?.exponent,
      coinGeckoId: currency.coingecko_id || undefined,
      coinImageUrl: currency.logo_URIs?.svg ?? currency.logo_URIs?.png
    };
  });

  const stakeCurrency: Currency =
    currencies.find((currency) => stakingDenoms.includes(currency.coinDenom)) ??
    currencies[0];

  const feeCurrencies: FeeCurrency[] = currencies
    // USE THE FEE DENOMS
    .filter((currency) => feeDenoms.includes(currency.coinMinimalDenom))
    .map((feeCurrency) => {
      if (!(feeCurrency.coinMinimalDenom in gasPriceSteps)) {
        return feeCurrency;
      }

      // has gas
      const gasPriceStep = gasPriceSteps[feeCurrency.coinMinimalDenom];
      return {
        ...feeCurrency,
        gasPriceStep
      };
    });

  const feeCurrenciesDefault: FeeCurrency[] = currencies
    // USE THE STAKE CURRENCY
    .filter((currency) => stakeCurrency.coinDenom === currency.coinDenom)
    .map((feeCurrency) => {
      if (!(feeCurrency.coinMinimalDenom in gasPriceSteps)) {
        return feeCurrency;
      }

      // has gas
      const gasPriceStep = gasPriceSteps[feeCurrency.coinMinimalDenom];
      return {
        ...feeCurrency,
        gasPriceStep
      };
    });

  const chainInfo: ChainInfo = {
    rpc: options.getRpcEndpoint(chain),
    rest: options.getRestEndpoint(chain),
    chainId: chain.chain_id,
    chainName: chain.pretty_name,
    bip44: {
      coinType: chain.slip44
    },
    bech32Config: Bech32Address.defaultBech32Config(chain.bech32_prefix),
    currencies: currencies,
    stakeCurrency: stakeCurrency || currencies[0],
    feeCurrencies:
      feeCurrencies.length !== 0 ? feeCurrencies : feeCurrenciesDefault,
    features
  };

  return chainInfo;
};
