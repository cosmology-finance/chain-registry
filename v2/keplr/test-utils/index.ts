import { AssetList, Chain, IBCInfo } from '@chain-registry/types';

import _assets from '../../../__fixtures__/assets.json';
import _chains from '../../../__fixtures__/chains.json';
import _ibc from '../../../__fixtures__/ibc.json';

export const assets: AssetList[] = _assets as AssetList[];
// TODO update fixtures, they may be missing properties and/or old schema
export const chains: Chain[] = _chains as any;

export const ibc: IBCInfo[] = _ibc as IBCInfo[];
