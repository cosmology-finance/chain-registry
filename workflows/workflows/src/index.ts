export * from './build';
export * from './convert';
export * from './registry';
export * from './registry-builder';
export * from './schema-typescript';
export * from './validator';
export { JSONSchema } from 'schema-typescript';
