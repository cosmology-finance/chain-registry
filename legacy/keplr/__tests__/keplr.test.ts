import { chainRegistryChainToKeplr } from '../src/';
import { assets, chains } from '../test-utils';

it('keplr', () => {
  const chain = chains.find((chain) => chain.chain_name === 'osmosis')!;
  const config = chainRegistryChainToKeplr(chain, assets);
  expect(config).toMatchSnapshot();
});

it('getExplorer', () => {
  const chain = chains.find((chain) => chain.chain_name === 'osmosis')!;
  const address = chain.apis!.rest![1]!.address!;

  const config = chainRegistryChainToKeplr(chain, assets, {
    getExplorer: () => 'https://myexplorer.com',
    getRestEndpoint: (_chain) => address
  });
  expect(config).toMatchSnapshot();
});
