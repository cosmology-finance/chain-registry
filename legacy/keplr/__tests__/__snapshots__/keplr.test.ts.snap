// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExplorer 1`] = `
{
  "bech32Config": {
    "bech32PrefixAccAddr": "osmo",
    "bech32PrefixAccPub": "osmopub",
    "bech32PrefixConsAddr": "osmovalcons",
    "bech32PrefixConsPub": "osmovalconspub",
    "bech32PrefixValAddr": "osmovaloper",
    "bech32PrefixValPub": "osmovaloperpub",
  },
  "bip44": {
    "coinType": 118,
  },
  "chainId": "osmosis-1",
  "chainName": "Osmosis",
  "currencies": [
    {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
      "coinMinimalDenom": "uosmo",
    },
    {
      "coinDecimals": 6,
      "coinDenom": "ION",
      "coinGeckoId": "ion",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/ion.svg",
      "coinMinimalDenom": "uion",
    },
    {
      "coinDecimals": 6,
      "coinDenom": "JKL",
      "coinGeckoId": "jackal-protocol",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/jackal/images/jkl.svg",
      "coinMinimalDenom": "ibc/8E697BDABE97ACE8773C6DF7402B2D1D5104DD1EEABE12608E3469B7F64C15BA",
    },
  ],
  "features": [
    "stargate",
    "no-legacy-stdTx",
    "ibc-transfer",
    "ibc-go",
    "cosmwasm",
    "wasmd_0.24+",
  ],
  "feeCurrencies": [
    {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
      "coinMinimalDenom": "uosmo",
      "gasPriceStep": {
        "average": 0.025,
        "high": 0.04,
        "low": 0,
      },
    },
  ],
  "rest": "https://lcd-osmosis.whispernode.com",
  "rpc": "https://osmosis-rpc.quickapi.com:443",
  "stakeCurrency": {
    "coinDecimals": 6,
    "coinDenom": "OSMO",
    "coinGeckoId": "osmosis",
    "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
    "coinMinimalDenom": "uosmo",
  },
}
`;

exports[`keplr 1`] = `
{
  "bech32Config": {
    "bech32PrefixAccAddr": "osmo",
    "bech32PrefixAccPub": "osmopub",
    "bech32PrefixConsAddr": "osmovalcons",
    "bech32PrefixConsPub": "osmovalconspub",
    "bech32PrefixValAddr": "osmovaloper",
    "bech32PrefixValPub": "osmovaloperpub",
  },
  "bip44": {
    "coinType": 118,
  },
  "chainId": "osmosis-1",
  "chainName": "Osmosis",
  "currencies": [
    {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
      "coinMinimalDenom": "uosmo",
    },
    {
      "coinDecimals": 6,
      "coinDenom": "ION",
      "coinGeckoId": "ion",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/ion.svg",
      "coinMinimalDenom": "uion",
    },
    {
      "coinDecimals": 6,
      "coinDenom": "JKL",
      "coinGeckoId": "jackal-protocol",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/jackal/images/jkl.svg",
      "coinMinimalDenom": "ibc/8E697BDABE97ACE8773C6DF7402B2D1D5104DD1EEABE12608E3469B7F64C15BA",
    },
  ],
  "features": [
    "stargate",
    "no-legacy-stdTx",
    "ibc-transfer",
    "ibc-go",
    "cosmwasm",
    "wasmd_0.24+",
  ],
  "feeCurrencies": [
    {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
      "coinMinimalDenom": "uosmo",
      "gasPriceStep": {
        "average": 0.025,
        "high": 0.04,
        "low": 0,
      },
    },
  ],
  "rest": "https://osmosis-lcd.quickapi.com:443",
  "rpc": "https://osmosis-rpc.quickapi.com:443",
  "stakeCurrency": {
    "coinDecimals": 6,
    "coinDenom": "OSMO",
    "coinGeckoId": "osmosis",
    "coinImageUrl": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
    "coinMinimalDenom": "uosmo",
  },
}
`;
